taichi/program/async_engine.cpp

```diff
@@ -205,11 +205,29 @@ void AsyncEngine::launch(Kernel *kernel, Context &context) {
     TaskLaunchRecord rec(context, kernel, kmeta.ir_handle_cached[i]);
     records.push_back(rec);
   }
-  sfg->insert_tasks(records, program->config.async_listgen_fast_filtering);
+  const auto &config = program->config;
+  sfg->insert_tasks(records, config.async_listgen_fast_filtering);
+  if ((config.async_flush_every > 0) &&
+      (sfg->num_pending_tasks() >= config.async_flush_every)) {
+    TI_TRACE("Async flushing {} tasks", sfg->num_pending_tasks());
+    flush();
+  }
 }
 
 void AsyncEngine::synchronize() {
   TI_AUTO_PROF;
+  flush();
+  queue.synchronize();
+
+  sync_counter_++;
+  // Clear SFG debug stats
+  cur_sync_sfg_debug_counter_ = 0;
+  cur_sync_sfg_debug_per_stage_counts_.clear();
```

Comment:
nit: When debugging SFG, I think the counter of #flushes is more useful than #syncs. Maybe include both #flushes and #syncs in the generated .dot file name?

Reply:
Yep, done